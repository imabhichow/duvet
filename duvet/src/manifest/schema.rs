use crate::Result;
use duvet_core::{fs::PathId, manifests::Builder, Fs};
use std::path::Path;

pub mod v1;

pub enum Schema {
    V1(v1::Schema),
}

impl Schema {
    pub fn parse(file: &Path, contents: &str) -> Result<Self> {
        match file.extension().and_then(|ext| ext.to_str()) {
            Some("toml") => {
                // TODO add version entry
                let manifest = toml::from_str(contents)?;
                Ok(Self::V1(manifest))
            }
            ext => unimplemented!("{:?}", ext),
        }
    }

    pub fn load(&self, fs: &Fs, path_id: PathId, manifest: &mut Builder) {
        match self {
            Self::V1(schema) => schema.load(fs, path_id, manifest),
        }
    }
}
