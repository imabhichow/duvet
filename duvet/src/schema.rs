use byteorder::BigEndian as BE;
use core::convert::TryInto;
use hash_hasher::HashedMap;
use sha2::{Digest, Sha256};
use zerocopy::{byteorder::U32, AsBytes, FromBytes, Unaligned};

#[derive(Default)]
struct Hasher(Sha256);

impl core::hash::Hasher for Hasher {
    fn write(&mut self, bytes: &[u8]) {
        self.0.update(bytes);
    }

    fn finish(&self) -> u64 {
        panic!("finish should not be called")
    }
}

macro_rules! id {
    ($name:ident) => {
        #[derive(AsBytes, FromBytes, Unaligned, Clone, Copy, Debug, PartialEq, Eq, Hash)]
        #[repr(C)]
        pub struct $name(pub(crate) U32<BE>);

        impl $name {
            pub fn new(value: u32) -> Self {
                Self(U32::new(value))
            }
        }

        impl From<u32> for $name {
            fn from(value: u32) -> Self {
                Self::new(value)
            }
        }

        impl From<U32<BE>> for $name {
            fn from(value: U32<BE>) -> Self {
                Self(value)
            }
        }

        impl PartialOrd for $name {
            fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
                Some(self.cmp(&other))
            }
        }

        impl Ord for $name {
            fn cmp(&self, other: &Self) -> core::cmp::Ordering {
                self.0.get().cmp(&other.0.get())
            }
        }
    };
}

id!(SourceId);
id!(SourceLocationId);
id!(TypeId);
id!(AnnotationId);

pub type Id = u128;
pub type ExpansionId = Id;
pub type NotificationId = Id;
pub type InstantiationId = Id;
pub type FileId = Id;

#[derive(Clone, Debug, Default)]
pub struct Report {
    pub instantiations: Set<Instantiation>,
    pub annotations: Set<Annotation>,
    pub types: Set<Type>,
    pub notifications: Set<Notification>,
    pub files: Set<File>,
    pub aliases: Set<Alias>,
}

#[derive(Clone, Debug)]
pub struct Set<T: SetEntry>(HashedMap<Id, T>);

impl<T: SetEntry> Default for Set<T> {
    fn default() -> Self {
        Self(HashedMap::default())
    }
}

pub trait SetEntry: core::hash::Hash {
    fn merge(&mut self, other: Self);
}

impl<T: SetEntry> Set<T> {
    pub fn insert(&mut self, value: T) -> Id {
        let mut hasher = Hasher::default();
        value.hash(&mut hasher);
        let id = hasher.0.finalize();
        let id = (id[..core::mem::size_of::<u128>()]).try_into().unwrap();
        let id = u128::from_le_bytes(id);
        self.insert_kv(id, value);
        id
    }

    fn insert_kv(&mut self, id: Id, value: T) {
        use std::collections::hash_map::Entry;

        match self.0.entry(id) {
            Entry::Occupied(mut entry) => {
                entry.get_mut().merge(value);
            }
            Entry::Vacant(entry) => {
                entry.insert(value);
            }
        }
    }

    pub fn get(&self, id: &Id) -> Option<&T> {
        self.0.get(id)
    }

    pub fn merge(&mut self, mut other: Self) {
        for (key, value) in other.0.drain() {
            self.insert_kv(key, value);
        }
    }
}

impl Report {
    pub fn merge(&mut self, other: Report) {
        self.instantiations.merge(other.instantiations);
        self.annotations.merge(other.annotations);
        self.types.merge(other.types);
        self.notifications.merge(other.notifications);
        self.files.merge(other.files);
        self.aliases.merge(other.aliases);
    }
}

/// Marks a region of text as instantiated
#[derive(Clone, Debug, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Instantiation {
    /// Optional name of the instantiation
    pub name: Option<String>,
    /// Target bytes of the instantiation
    pub range: Byterange,
    /// Target file of the instantiation
    pub file: FileId,
}

impl SetEntry for Instantiation {
    fn merge(&mut self, other: Self) {
        check_collision(self, &other)
    }
}

/// Indicates that several files actually point to the same one
#[derive(Clone, Debug, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Alias {
    /// Target file
    pub file: FileId,
    /// A list of aliases that point to the target file
    pub aliases: Vec<FileId>,
}

impl SetEntry for Alias {
    fn merge(&mut self, other: Self) {
        check_collision(self, &other)
    }
}

/// General file information
#[derive(Clone, Debug, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct File {
    /// Optional name for the file. This will default to the basename
    pub name: Option<String>,
    /// Path to the file
    pub path: String,
    /// Language of the file.
    pub language: Option<String>,
}

impl SetEntry for File {
    fn merge(&mut self, other: Self) {
        check_collision(self, &other)
    }
}

/// Annotates a section of text
#[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Annotation {
    /// The type of the annotation
    pub type_id: TypeId,
}

impl SetEntry for Annotation {
    fn merge(&mut self, other: Self) {
        check_collision(self, &other)
    }
}

/// Associates an instantiation with a subset of bytes
#[derive(Clone, Debug, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Region {
    /// The target instantiation
    pub instantiation: InstantiationId,
    /// The subset of bytes
    pub range: Byterange,
}

/// Defines the charactaristics of a particular type of annotation
#[derive(Clone, Debug, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Type {
    /// The name of the type
    pub name: String,
    /// The type will be fulfilled if all of the dependencies are
    pub transparent: bool,
    /// A list of types that this type requires
    pub dependencies: Vec<Dependency>,
}

impl SetEntry for Type {
    fn merge(&mut self, other: Self) {
        check_collision(self, &other)
    }
}

/// A dependency of an annotation type
#[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Dependency {
    /// Include a notification if this dependency is not met
    pub notification: Option<NotificationId>,
    /// Include a description of the dependency
    pub description: Option<String>,
    /// An expression of annotation types that must be met
    pub target: DependencyTarget,
}

/// A target expression for dependencies
#[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub enum DependencyTarget {
    /// A single type
    Id(TypeId),
    /// If any of the types are met, the dependency is ok
    Any(Vec<DependencyTarget>),
    /// If all of the types are met, the dependency is ok
    All(Vec<DependencyTarget>),
}

/// A notification that is displayed when dependencies are not met
#[derive(Clone, Debug, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Notification {
    /// The title of the notification
    pub title: Option<String>,
    /// The description of the notification
    pub message: Option<String>,
    /// The severity level of the notification
    pub level: Level,
}

impl SetEntry for Notification {
    fn merge(&mut self, other: Self) {
        check_collision(self, &other)
    }
}

/// A non-exlusive single region of bytes
#[derive(Clone, Copy, Debug, Default, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Byterange {
    /// Start offset
    start: usize,
    /// End offset
    end: usize,
}

/// A severity level for a notification
#[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub enum Level {
    Fatal,
    Error,
    Warning,
    Info,
    Debug,
}

impl Default for Level {
    fn default() -> Self {
        Self::Info
    }
}

fn check_collision<T: PartialEq>(a: &T, b: &T) {
    if cfg!(debug_assertions) && a != b {
        panic!("hash collision detected!");
    }
}
