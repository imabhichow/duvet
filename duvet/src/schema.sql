BEGIN;

CREATE TABLE IF NOT EXISTS sources (
    id INTEGER PRIMARY KEY,
    hash BLOB UNIQUE NOT NULL
);

CREATE TABLE IF NOT EXISTS source_locations (
    id INTEGER PRIMARY KEY,
    location VARCHAR UNIQUE,
    source_id INTEGER NOT NULL,
    title VARCHAR
);

CREATE TABLE IF NOT EXISTS source_lines (
    source_id INTEGER,
    line INTEGER,
    columns INTEGER,
    offset INTEGER,
    PRIMARY KEY(source_id, line)
);

CREATE TABLE IF NOT EXISTS instances (
    id INTEGER PRIMARY KEY,
    name VARCHAR NOT NULL,
    source_location_id INTEGER NOT NULL,
    start_offset INTEGER NOT NULL,
    end_offset INTEGER NOT NULL
);

CREATE TABLE IF NOT EXISTS annotations (
    id INTEGER PRIMARY KEY,
    type_id INTEGER NOT NULL
);

CREATE TABLE IF NOT EXISTS annotation_labels (
    annotation_id INTEGER,
    label VARCHAR,
    PRIMARY KEY(annotation_id, label)
);

CREATE TABLE IF NOT EXISTS annotation_source_regions (
    annotation_id INTEGER NOT NULL,
    start_offset INTEGER NOT NULL,
    end_offset INTEGER NOT NULL,
    source_location_id INTEGER NOT NULL,
    PRIMARY KEY(annotation_id, start_offset, end_offset, source_location_id)
);

CREATE TABLE IF NOT EXISTS annotation_instance_regions (
    annotation_id INTEGER NOT NULL,
    start_offset INTEGER NOT NULL,
    end_offset INTEGER NOT NULL,
    instance_id INTEGER NOT NULL,
    PRIMARY KEY(annotation_id, start_offset, end_offset, instance_id)
);

CREATE TABLE IF NOT EXISTS annotation_relations (
    source_id INTEGER NOT NULL,
    target_id INTEGER NOT NULL,
    type_id INTEGER NOT NULL,
    PRIMARY KEY(source_id, target_id, type_id)
);

CREATE TABLE IF NOT EXISTS annotation_metrics (
    annotation_id INTEGER NOT NULL,
    name VARCHAR NOT NULL,
    value INTEGER NOT NULL,
    PRIMARY KEY(annotation_id, name)
);

END;
