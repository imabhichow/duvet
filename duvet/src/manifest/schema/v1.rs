use duvet_core::{fs::PathId, manifests::Builder, Fs};
use serde::{de::Deserializer, Deserialize};

#[derive(Debug, Deserialize)]
pub struct Schema {
    pub compliance: Compliance,
}

impl Schema {
    pub fn load(&self, fs: &Fs, path_id: PathId, manifest: &mut Builder) {
        // TODO
    }
}

pub use compliance::Compliance;

pub mod compliance {
    use super::*;

    #[derive(Debug, Deserialize)]
    pub struct Compliance {
        #[serde(rename = "source")]
        pub sources: Vec<Source>,

        #[serde(rename = "requirement")]
        pub requirements: Vec<Requirement>,
    }

    #[derive(Debug, Deserialize)]
    pub struct Source {
        pub patterns: Vec<Pattern>,
        #[serde(rename = "comment-style")]
        pub comment_style: CommentStyle,
    }

    #[derive(Debug, Deserialize)]
    pub struct CommentStyle {
        pub meta: String,
        pub content: String,
    }

    #[derive(Debug, Deserialize)]
    pub struct Requirement {
        pub patterns: Vec<Pattern>,
    }

    #[derive(Debug, Deserialize)]
    pub struct Spec {
        pub markdown: Vec<Markdown>,
        pub ietf: Vec<Ietf>,
    }

    #[derive(Debug, Deserialize)]
    pub struct Markdown {
        pub patterns: Vec<Pattern>,
    }

    #[derive(Debug, Deserialize)]
    pub struct Ietf {
        pub id: String,
        pub title: Option<String>,
        pub url: String,
        #[serde(default)]
        pub aliases: Vec<String>,
    }
}

#[derive(Debug)]
pub struct Pattern(globset::Glob);

impl<'de> Deserialize<'de> for Pattern {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let value = String::deserialize(deserializer)?;
        let value = globset::Glob::new(&value).map_err(serde::de::Error::custom)?;
        Ok(Self(value))
    }
}
