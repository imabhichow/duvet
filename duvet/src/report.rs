use crate::schema::AnnotationId;
use byteorder::BigEndian as BE;
use core::{convert::TryInto, fmt, ops::Range};
use sled::{IVec, Result, Tree};
use std::collections::HashMap;
use zerocopy::{AsBytes, FromBytes, LayoutVerified, Unaligned, U32};

#[repr(C)]
#[derive(AsBytes, FromBytes, Unaligned, Clone, Copy, Debug)]
struct AnnotationEntry {
    id: AnnotationId,
    end: U32<BE>,
}

pub struct AnnotationSet {
    annotations: Tree,
}

impl AnnotationSet {
    pub fn new(annotations: Tree) -> Self {
        annotations.set_merge_operator(Self::merge_entry);

        Self { annotations }
    }

    pub fn insert(&mut self, id: AnnotationId, bytes: Range<u32>) -> Result<()> {
        let start = bytes.start;
        let end = bytes.end;

        if start >= end {
            return Ok(());
        }

        let entry = AnnotationEntry {
            id,
            end: U32::new(end),
        };

        self.annotations
            .merge(bytes.start.to_be_bytes(), entry.as_bytes())?;
        self.annotations
            .merge(bytes.end.to_be_bytes(), entry.as_bytes())?;

        Ok(())
    }

    pub fn finish(self, out: &Tree) -> Result<()> {
        let mut current: HashMap<AnnotationId, u32> = HashMap::new();
        let mut value_buf = vec![];

        for entry in self.annotations.iter() {
            let (key, value) = entry?;
            let offset = get_offset(&key);

            let entries =
                <LayoutVerified<_, [AnnotationEntry]>>::new_slice_unaligned(value.as_ref())
                    .unwrap()
                    .into_slice();

            for entry in entries {
                let end = entry.end.get();

                current
                    .entry(entry.id)
                    .and_modify(|prev| *prev = (*prev).max(end))
                    .or_insert(end);
            }

            // delete all of the annotations that ended on this entry
            current.retain(|anno_id, end| {
                if *end > offset {
                    value_buf.extend_from_slice(anno_id.as_bytes());
                    true
                } else {
                    false
                }
            });

            out.insert(key, value_buf.as_slice())?;

            value_buf.clear();
        }

        Ok(())
    }

    #[allow(clippy::unnecessary_wraps)]
    fn merge_entry(_key: &[u8], old_value: Option<&[u8]>, merged_bytes: &[u8]) -> Option<Vec<u8>> {
        let mut value = old_value
            .map(Vec::from)
            .unwrap_or_else(|| Vec::with_capacity(merged_bytes.len()));
        value.extend_from_slice(merged_bytes);
        Some(value)
    }
}

fn get_offset<V: AsRef<[u8]>>(key: V) -> u32 {
    u32::from_be_bytes(key.as_ref().try_into().unwrap())
}

struct Regions(Tree);

impl Regions {
    pub fn new(regions: Tree) -> Self {
        Self(regions)
    }

    pub fn iter(&self) -> impl Iterator<Item = Result<Region>> + '_ {
        self.0.iter().map(|v| v.map(Region::new))
    }
}

impl fmt::Debug for Regions {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_list()
            .entries(self.iter().filter_map(|r| r.ok()))
            .finish()
    }
}

struct Region {
    pub offset: u32,
    ids: IVec,
}

impl fmt::Debug for Region {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        f.debug_struct("Region")
            .field("offset", &self.offset)
            .field("ids", &self.ids())
            .finish()
    }
}

impl Region {
    pub fn new((offset, ids): (IVec, IVec)) -> Self {
        let offset = get_offset(offset);
        Self { offset, ids }
    }

    pub fn ids(&self) -> &[AnnotationId] {
        <LayoutVerified<_, [AnnotationId]>>::new_slice_unaligned(self.ids.as_ref())
            .unwrap()
            .into_slice()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn regions(regions: &[(u32, Range<u32>)]) -> Regions {
        let db = sled::Config::new().temporary(true).open().unwrap();
        let annos = db.open_tree("annos").unwrap();

        let mut set = AnnotationSet::new(annos);

        for (id, bytes) in regions.iter().cloned() {
            set.insert(AnnotationId::new(id), bytes).unwrap();
        }

        let regions = db.open_tree("regions").unwrap();

        set.finish(&regions).unwrap();

        Regions::new(regions)
    }

    #[test]
    fn checked() {
        bolero::check!()
            .with_type::<Vec<(_, Range<u32>)>>()
            .for_each(|entries| {
                let regions = regions(entries);

                let mut last_count = 0;
                for entry in regions.iter() {
                    let region = entry.unwrap();
                    if last_count == 0 {
                        assert!(!region.ids().is_empty(), "empty regions should not repeat");
                    }
                    last_count = region.ids().len();
                }
            });
    }

    #[test]
    fn overlap() {
        insta::assert_debug_snapshot!(regions(&[(1, 0..2), (2, 2..4), (3, 0..5)]));
    }
}
