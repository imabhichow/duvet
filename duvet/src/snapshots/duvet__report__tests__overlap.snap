---
source: duvet/src/report.rs
expression: "regions(&[(1, 0..2), (2, 2..4), (3, 0..5)])"

---
[
    Region {
        offset: 0,
        ids: [
            AnnotationId(
                U32(1),
            ),
            AnnotationId(
                U32(3),
            ),
        ],
    },
    Region {
        offset: 2,
        ids: [
            AnnotationId(
                U32(2),
            ),
            AnnotationId(
                U32(3),
            ),
        ],
    },
    Region {
        offset: 4,
        ids: [
            AnnotationId(
                U32(3),
            ),
        ],
    },
    Region {
        offset: 5,
        ids: [],
    },
]
